pkg/virt-launcher/virtwrap/api/converter.go

```diff
@@ -1014,27 +1010,64 @@ func Convert_v1_VirtualMachine_To_api_Domain(vmi *v1.VirtualMachineInstance, dom
 	return nil
 }
 
-func calculateRequestedVCPUs(vmi *v1.VirtualMachineInstance) uint32 {
-	cores := uint32(0)
-	if vmi.Spec.Domain.CPU != nil && vmi.Spec.Domain.CPU.Cores != 0 {
+func getCPUTopology(vmi *v1.VirtualMachineInstance) *CPUTopology {
+	return &CPUTopology{
+		Sockets: calculateRequestedSockets(vmi),
+		Cores:   calculateRequestedCPUs(vmi),
+		Threads: calculateRequestedThreads(vmi),
+	}
+}
+
+func calculateRequestedSockets(vmi *v1.VirtualMachineInstance) uint32 {
+	sockets := uint32(1)
+	cpu := vmi.Spec.Domain.CPU
+	if cpu.Sockets != 0 {
+		return cpu.Sockets
+	}
+
+	if cpu.Cores == 0 && cpu.Threads == 0 {
+		//if cores, sockets, threads are not set, take value from domain resources request or limits and
+		//set value into sockets, which have best performance (https://bugzilla.redhat.com/show_bug.cgi?id=1653453)
+		resources := vmi.Spec.Domain.Resources
+		if cpuRequests, ok := resources.Requests[k8sv1.ResourceCPU]; ok {
```

Comment:
CPU Resource can also be in milli format aka `0.100` or alike, thus no int.

But anything else than int is not helpful for number of sockets.

Reply:
It would only be an int in the guaranteed case.

Reply:
Did you notice this is the original code just moved to different place?

Reply:
But you are right, we need to round the value up to make sure it is a greater than zero  integer.

Reply:
If i am correct, cpuRequests.Value() should return round int64 number.

Reply:
We were digging into kubernetes source with @vladikr and it seems that there are two possible implementations, [one](https://github.com/kubernetes/apimachinery/blob/master/pkg/api/resource/quantity.go#L684) just strips the fractional part (0.5 -> 0) and the [other one](https://github.com/golang/build/blob/master/kubernetes/api/quantity.go#L356) rounds up.